perf: (googlecalendar): batch freebusy calls by delegation credential

volnei opened the pull request:
## What does this PR do?

This PR optimizes Google Calendar API calls by batching `getFreeBusyResult` requests more efficiently. Instead of making individual API calls for each calendar, it:

1. **Groups calendars by delegation credential** - Calendars with the same `delegationCredentialId` are batched together
2. **Chunks groups to respect API limits** - Each delegation credential group is split into chunks of maximum 50 calendars (Google's API limit)
3. **Processes groups in parallel** - Uses `Promise.allSettled()` to fetch availability for all groups/chunks concurrently with graceful partial failure handling

**Architecture:**

- `CalendarBatchService` - Simple utility class that determines which calendar types support batching (currently only `google_calendar`)
- `CalendarBatchWrapper` - Decorator that wraps the original calendar service and optimizes `getAvailability()` calls
- `getCalendar()` integration - Decision flow: cache wrapper (if enabled) → batch wrapper → raw calendar

## Updates since last revision

- **Reimplemented batching as a wrapper pattern** - Moved batching logic from `CalendarService` to `CalendarBatchWrapper` for cleaner separation of concerns
- **Added comprehensive README** in `packages/features/calendar-batch/` explaining how the feature works
- **Added CalendarBatchWrapper tests** covering:
    - Separate calls for calendars without `delegationCredentialId`
    - Batching calendars with same `delegationCredentialId`
    - Chunking into groups of 50 for API limits
    - Mixed calendars handling
    - `fallbackToPrimary` with empty array
    - Result flattening from batched calls
    - Pass-through method delegation
- **Fixed cubic-dev-ai review comments** - Removed misleading comment, fixed typo in `getCalendar.ts`
- **Fixed** **`CalendarBatchWrapper.getAvailability`** **signature** - Added missing `shouldServeCache` parameter to match the `Calendar` interface. This was causing the 4th argument to be misinterpreted.
- **Added integration test for call count verification** - New test in `CalendarService.test.ts` that verifies `CalendarBatchWrapper` makes separate API calls for different `delegationCredentialIds`
- **Added** **`shouldServeCache`** **forwarding test** - Explicit test to verify the `shouldServeCache` parameter is correctly passed through to the original calendar (guards against past regression)
- **Added order-independent batching verification test** - Verifies calendar grouping by checking set membership rather than call order, ensuring robust test coverage
- **Refactored tests to be order-independent** - Addressed Sean's review comment about `Promise.all` not guaranteeing execution order. Tests now use set comparisons instead of `toHaveBeenNthCalledWith` to avoid potential flakiness.
- **Added comprehensive tests for `resolveCalendarServeStrategy()`** - New test file at `packages/app-store/_utils/__tests__/getCalendar.test.ts` with 15 tests covering:
    - Cache wrapper selection when cache is supported and enabled
    - Batch wrapper selection when cache is not supported but batch is supported
    - Cache wrapper priority over batch wrapper when both are supported
    - Original calendar fallback when neither optimization is supported
    - Feature flag determination when `shouldServeCache` is undefined
    - Feature flag disabled globally / user-level scenarios
    - Integration with `CalendarBatchService.isSupported` for different credential types
    - Edge cases (null credential, missing key, credential type suffixes)
- **Changed `Promise.all` to `Promise.allSettled` for partial failure handling** - Both `getAvailability()` and `getAvailabilityWithTimeZones()` now use `Promise.allSettled()` instead of `Promise.all()`. This means:
    - If some batches fail, successful batches still return their results
    - Failed batches are logged with `log.warn` including error details
    - If all batches fail, an empty array is returned instead of throwing
- **Added partial failure tests** - 3 new tests in `CalendarBatchWrapper.test.ts`:
    - Returns partial results when some batches fail
    - Returns empty array when all batches fail
    - Does not throw when some batches fail
- **Added comprehensive tests for `getAvailabilityWithTimeZones()`** - 9 new tests covering:
    - Batching behavior (separate calls, batching by delegationCredentialId, chunking to 50)
    - Partial failure handling (partial results, all fail, no throw)
    - Result flattening and combining
    - Edge case for underlying calendar not implementing the method

## Human Review Checklist

Please pay special attention to these areas:

- [ ] **Partial failure behavior**: Verify that returning partial results (instead of failing entirely) is acceptable. Previously, if one batch failed, the entire operation would reject. Now it returns whatever succeeded.
- [ ] **"All fail" behavior**: If ALL batches fail, an empty array is returned. This could be interpreted as "fully available" - verify this is acceptable or if throwing would be safer.
- [ ] **Error logging sufficiency**: Failed batches are logged with `log.warn` - verify this provides enough visibility for debugging production issues
- [ ] **Wrapper delegation**: Verify `CalendarBatchWrapper` correctly delegates all non-availability methods to the original calendar
- [ ] **Credential splitting logic**: `splitCalendars()` separates calendars by `delegationCredentialId` - calendars without it are processed one-by-one (conservative approach)
- [ ] **Empty calendar fallback**: When no calendars provided, a single call with `[]` is made so `fallbackToPrimary` can be honored
- [ ] **getCalendar decision flow**: Cache wrapper takes precedence over batch wrapper - verify this is the intended behavior
- [ ] **Signature fix**: Verify `shouldServeCache` parameter is correctly forwarded through `CalendarBatchWrapper.getAvailability` to the original calendar
- [ ] **Type casts**: Review `as unknown as Calendar` casts in `getCalendar.ts` to ensure they don't hide type issues
- [ ] **New test mocks**: Verify the mocks in `getCalendar.test.ts` accurately represent real service behavior

## How should this be tested?

**Environment Setup:**

- Google Calendar integration configured with delegation credentials
- Multiple calendars across different delegation credentials (>50 total to test chunking)

**Test Scenarios:**

1. **Basic grouping**: Calendars with same `delegationCredentialId` should be batched in single API call
2. **Chunking behavior**: Groups >50 calendars should be split into multiple API calls of max 50 each
3. **Mixed credentials**: Calendars with different delegation credentials should be processed in separate batches
4. **Fallback logic**: When no calendars selected, should still fallback to primary calendar correctly
5. **Partial failure handling**: If one delegation credential group fails (e.g., 403 permission error), other groups should still return their results

**Automated Tests:**
- Run `TZ=UTC yarn test packages/features/calendar-batch` - the CalendarBatchWrapper tests should pass (24 tests)
- Run `TZ=UTC yarn test packages/app-store/_utils/__tests__/getCalendar.test.ts` - the resolveCalendarServeStrategy tests should pass (15 tests)

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox. Added README in `packages/features/calendar-batch/`
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

---

**Link to Devin run**: https://app.devin.ai/sessions/86263a1be28640e9aca91ed6c69a3c25
**Previous Devin sessions**: https://app.devin.ai/sessions/fad959b554064be3b0eff1d4766cd4f3, https://app.devin.ai/sessions/e64091b2ff88429dbbcd9a4054955a06, https://app.devin.ai/sessions/c6bc9002033449b9bdfd9798e3fe1f33
**Requested by**: keith@cal.com (keith@cal.com) (@keithwillcode), Volnei Munhoz (@volnei)

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

devin-ai-integration[bot] committed:
perf(googlecalendar): batch freebusy calls by delegation credential

- Group selectedCalendars by delegationCredentialId before making API calls
- Make one batched freebusy query per delegation credential group
- Reduces total API calls while respecting credential boundaries
- Maintains existing caching behavior per group
- Updated both getAvailability and getAvailabilityWithTimeZones methods
- Added groupCalendarsByDelegationCredential helper method
- Handles edge case when no calendars provided but fallbackToPrimary is true
- Fixed linting issue: replaced hasOwnProperty with 'in' operator

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
test(googlecalendar): add test for delegation credential batching

- Verify calendars are grouped by delegationCredentialId
- Ensure exactly 3 API calls made for 3 delegation credential groups
- Confirm all busy times from different groups are properly returned
- Fix type-safety issues by replacing 'as any' with proper type constraints
- Fix ESLint warnings: unused variables and any types in mock functions

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
fix(googlecalendar): ensure fallback logic works with empty calendar groups

- Handle empty calendar groups by ensuring at least one iteration
- Add test for chunking groups larger than 50 calendars
- Verify all delegation credential batching logic works correctly

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

volnei committed:
Merge branch 'main' into devin/batch-google-calendar-by-delegation-1759855804

devin-ai-integration[bot] committed:
fix: resolve merge conflicts with main branch

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
fix: fix remaining type errors from merge conflict resolution

- Changed getCacheOrFetchAvailability to getFreeBusyData in getAvailabilityWithTimeZones
- Removed orphaned merge conflict marker in test file

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
fix: reset test file to original PR version and update method name

- Reset CalendarService.test.ts to original PR version (0e9eb9e97a)
- Updated getCacheOrFetchAvailability to getFreeBusyData to match main branch

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
fix: reset test file to main branch version

The original PR's test file had tests for caching features that have been
removed from main. Reset to main's version to fix type errors.

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

volnei commented on packages/app-store/googlecalendar/lib/CalendarService.ts:
```diff
@@ -531,60 +565,73 @@ export default class GoogleCalendarService implements Calendar {
-          timeZone: timeZoneMap.get(freeBusy.id) || "",
-        };
-      });
+      const allResults = await Promise.all(
```
@devin does it makes sense?

volnei commented on packages/app-store/googlecalendar/lib/CalendarService.ts:
```diff
@@ -748,32 +795,51 @@ export default class GoogleCalendarService implements Calendar {
+
+            this.log.debug(
+              "[Cache Miss] Fetching for delegation credential group chunk",
+              safeStringify({ delegationCredentialId, calendarCount: calendarIdsChunk.length })
```
@devin does it makes sense?

volnei committed:
Merge branch 'main' into devin/batch-google-calendar-by-delegation-1759855804

volnei committed:
Merge branch 'main' into devin/batch-google-calendar-by-delegation-1759855804

devin-ai-integration[bot] committed:
test(googlecalendar): add tests for delegation credential batching logic

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

volnei committed:
Merge branch 'main' into devin/batch-google-calendar-by-delegation-1759855804

volnei committed:
Merge branch 'main' into devin/batch-google-calendar-by-delegation-1759855804

volnei committed:
Revamp devin's implementation

volnei committed:
Remove comsoles.log

devin-ai-integration[bot] committed:
fix(tests): update delegation credential batching tests to match reimplementation

- Remove tests for private methods that no longer exist (groupCalendarsByDelegationCredential, chunkArray)
- Update getAvailability test to verify calendar fetching without expecting multiple API calls per delegation credential
- Keep existing tests for fallback to primary calendar and non-google calendar handling

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

volnei committed:
Remove dead code

volnei committed:
Merge branch 'devin/batch-google-calendar-by-delegation-1759855804' of github.com:calcom/cal.com into devin/batch-google-calendar-by-delegation-1759855804

volnei committed:
improve documentation

devin-ai-integration[bot] committed:
docs: add README explaining Google Calendar availability batching feature

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
docs: translate README to English

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
docs: move README to calendar-batch package with comprehensive documentation

- Remove README from googlecalendar lib (wrong location)
- Add comprehensive README to packages/features/calendar-batch/
- Document CalendarBatchService and CalendarBatchWrapper
- Explain how getCalendar() integrates with batching
- Include architecture, data model, and performance considerations

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
fix: address cubic-dev-ai review comments

- Remove misleading comment from getCalendar.ts cache block
- Fix typo 'optmization' -> 'optimization' in comment
- Add comprehensive tests for CalendarBatchWrapper batching behavior
  - Test separate calls for calendars without delegationCredentialId
  - Test batching calendars with same delegationCredentialId
  - Test chunking into groups of 50 for API limits
  - Test mixed calendars handling
  - Test fallbackToPrimary with empty array
  - Test result flattening from batched calls
  - Test pass-through methods delegation

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

volnei committed:
Merge branch 'devin/batch-google-calendar-by-delegation-1759855804' of github.com:calcom/cal.com into devin/batch-google-calendar-by-delegation-1759855804

devin-ai-integration[bot] committed:
fix: add shouldServeCache param to CalendarBatchWrapper and verify batching call count

- Fix CalendarBatchWrapper.getAvailability signature to match Calendar interface (add shouldServeCache param)
- Update CalendarBatchWrapper tests to pass shouldServeCache parameter
- Add integration test in CalendarService.test.ts that verifies CalendarBatchWrapper makes separate API calls for different delegationCredentialIds (call count assertion)
- This fixes the getCalendarsEvents test failures caused by signature mismatch

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

volnei committed:
Improve CalendarBatchImplementation

devin-ai-integration[bot] committed:
test: add shouldServeCache forwarding and order-independent batching verification tests

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

sean-brydon commented on packages/features/calendar-batch/lib/CalendarBatchWrapper.ts:
```diff
@@ -0,0 +1,257 @@
+      );
+    }
+
+    const results = await Promise.all(tasks);
```
Should we use allSettled here? As this seems if one calendar fetch fails they all will.

volnei replied:
when it fails it will fallbackToPrimary... seems okay to me

volnei commented on packages/features/calendar-batch/lib/CalendarBatchWrapper.ts:
```diff
@@ -0,0 +1,257 @@
+        (sum, calendars) => sum + calendars.length,
+        0
+      ),
+    });
```
yes.. CalendarBatchWrapper only runs if it's a Google Calendar, locked by CalendarBatchService.

sean-brydon commented on packages/features/calendar-batch/lib/__tests__/CalendarBatchWrapper.test.ts:
```diff
@@ -0,0 +1,297 @@
+        [selectedCalendars[2]],
+        undefined,
+        false
+      );
```
Now i think about it - not sure if this applys when youre mocking and testing like this 🤔

volnei replied:
@sean-brydon order doesnt matter here once it was called =)

volnei committed:
Merge branch 'devin/batch-google-calendar-by-delegation-1759855804' of github.com:calcom/cal.com into devin/batch-google-calendar-by-delegation-1759855804

volnei committed:
Merge branch 'main' into devin/batch-google-calendar-by-delegation-1759855804

devin-ai-integration[bot] committed:
test: make CalendarBatchWrapper tests order-independent

Refactored tests to avoid relying on Promise.all execution order:
- 'should make separate calls for calendars without delegationCredentialId'
  now uses set comparison instead of toHaveBeenNthCalledWith
- 'should batch calendars with the same delegationCredentialId together'
  now finds calls by delegation credential instead of call order

This addresses Sean's review comment about potential flakiness due to
Promise.all not guaranteeing execution order of parallel promises.

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

keithwillcode reviewed (changes requested):
Overall looking very nice!

Left a comment regarding the hardcoded ID.

volnei committed:
Remove hardcoded ID

volnei commented:
I just removed the hard-coded and agreed to test it on qa first! 🚀

volnei committed:
Merge branch 'devin/batch-google-calendar-by-delegation-1759855804' of github.com:calcom/cal.com into devin/batch-google-calendar-by-delegation-1759855804

devin-ai-integration[bot] committed:
test: add comprehensive tests for resolveCalendarServeStrategy

Co-Authored-By: keith@cal.com <keithwillcode@gmail.com>

devin-ai-integration[bot] committed:
feat: use Promise.allSettled for partial failure handling in CalendarBatchWrapper

- Changed Promise.all to Promise.allSettled in getAvailability and getAvailabilityWithTimeZones
- Returns partial results when some batches fail instead of failing entirely
- Logs warnings for failed batches with error details
- Added tests for partial failure scenarios

Co-Authored-By: keith@cal.com <keithwillcode@gmail.com>

devin-ai-integration[bot] committed:
test: add comprehensive tests for getAvailabilityWithTimeZones

- Added batching behavior tests (separate calls, batching by delegationCredentialId, chunking)
- Added partial failure handling tests (partial results, all fail, no throw)
- Added edge case test for underlying calendar not implementing the method
- Total: 24 tests now covering both getAvailability and getAvailabilityWithTimeZones

Co-Authored-By: keith@cal.com <keithwillcode@gmail.com>

keithwillcode committed:
Merge branch 'main' into devin/batch-google-calendar-by-delegation-1759855804

volnei committed:
Merge branch 'main' into devin/batch-google-calendar-by-delegation-1759855804

devin-ai-integration[bot] committed:
fix: add required serviceAccountKey fields to delegatedTo mock objects

Co-Authored-By: keith@cal.com <keithwillcode@gmail.com>

devin-ai-integration[bot] committed:
fix: add missing client_id and private_key to serviceAccountKey mock

Co-Authored-By: keith@cal.com <keithwillcode@gmail.com>

keithwillcode reviewed (approved):
Let's get this tested on QA!
